app/Repository/Member/MemberMatch.php

```diff
@@ -51,11 +62,259 @@ class MemberMatch {
+			if ( count( $matches ) ) {
+				return self::create( $matches, false );
+			}
+		}
+		
+		// don't proceed, if we don't have first and last name
+		if ( ! ( $member->firstName->getValue() && $member->lastName->getValue() ) ) {
+			return new MemberMatch( self::NO_MATCH, [] );
+		}
+		
+		// search webling by first and last name
+		$query   = self::buildNameQuery( $member );
+		$matches = self::matchByQuery( $query, $rootGroups, $memberRepository );
+		
+		if ( ! empty( $matches ) ) {
+			// make sure we filter out all entries where only the beginning of the
+			// name was identical, but the given name was no a short name
+			self::removeWrongNameMatches( $matches, $member->firstName->getValue(),
+				$member->lastName->getValue() );
+		}
+		
+		if ( ! empty( $matches ) && $member->zip->getValue() ) {
+			// filter out all results, where the zip didn't match
+			self::removeWrongZipMatches( $matches, $member->zip->getValue() );
+			
+			return self::create( $matches, false );
+		}
+		
+		return self::create( $matches, true );
+	}
+	
+	/**
+	 * Return webling query to find members by email
+	 *
+	 * @param Member $member
+	 *
+	 * @return string
+	 */
+	private static function buildEmailQuery( Member $member ): string {
+		$query = [];
+		if ( $member->email1->getWeblingValue() ) {
+			$query[] = "`{$member->email1->getWeblingKey()}`,`{$member->email2->getWeblingKey()}` = '{$member->email1->getWeblingValue()}'";
+		}
+		
+		if ( $member->email2->getWeblingValue() ) {
+			$query[] = "`{$member->email1->getWeblingKey()}`,`{$member->email2->getWeblingKey()}` = '{$member->email2->getWeblingValue()}'";
+		}
+		
+		return implode( ' OR ', $query );
+	}
+	
+	/**
+	 * Return members that matched the given query and log exceptions that
+	 * should not occur here.
+	 *
+	 * @param string $query
+	 * @param array $rootGroups
+	 * @param MemberRepository $memberRepository
+	 *
+	 * @return Member[]
+	 *
+	 * @throws ClientException
+	 * @throws GroupNotFoundException
+	 * @throws WeblingAPIException
+	 */
+	private static function matchByQuery(
+		string $query,
+		array $rootGroups,
+		MemberRepository $memberRepository
+	): array {
+		try {
+			return $memberRepository->find( $query, $rootGroups );
+		} catch ( InvalidFixedValueException
+		| MemberNotFoundException
+		| MemberUnknownFieldException
+		| MultiSelectOverwriteException
+		| ValueTypeException
+		| WeblingFieldMappingConfigException $e ) {
+			Log::debug( $e->getFile() . ':' . $e->getLine() . "\n" . $e->getMessage() . $e->getTraceAsString(),
+				[ 'Query' => $query, 'Root Groups' => $rootGroups ] );
+			
+			return [];
+		}
+	}
+	
+	/**
+	 * Remove entries of the matches where the first name doesn't match
+	 *
+	 * The comparison is not case sensitive. If the matches don't contain a
+	 * first name, they will still be part of the result.
+	 *
+	 * @param Member[] $matches
+	 * @param string $firstName
+	 */
+	private static function selectByFirstName( array &$matches, string $firstName ) {
```

Comment:
Ev. gleiches Naming verwenden wie bei den weiteren Methoden mit ähnlicher Funktionalität.
Allenfalls auch hier beibehalten und die anderen umbenennen (selectByName, selectByZip). Sind die kürzeren Namen, aber auch weniger präzise, was genau passiert.
```suggestion
 	private static function removeWrongFirstNameMatches( array &$matches, string $firstName ) {
```

Reply:
Weil die Funktionen unterschiedlich sind, würde ich den Namen bewusst unterschiedlich wählen, oder? Siehe auch: https://github.com/grueneschweiz/weblingservice/pull/21#discussion_r246049824

Reply:
Ok. Ich finde zwar remove... intuitiver, was genau passiert, aber gleichzeitig gibt es Verwirrung, weil es nicht exakt das gleiche macht wie die anderen remove... Methoden.
Von dem her würde ich den Namen so lassen, ausser man würde den unten bei removeWrongNameMatches erwähnten bool-parameter einführen (IMHO aber nicht nötig).